diagnostics/diagnostics_manager.rs

```diff
@@ -0,0 +1,217 @@
+        pub fn submit_action_success(&self, database_name: Option<&str>, action_kind: ActionKind);
+        pub fn submit_action_fail(&self, database_name: Option<&str>, action_kind: ActionKind);
+        pub fn increment_load_count(&self, database_name: &str, connection_: LoadKind);
+        pub fn decrement_load_count(&self, database_name: &str, connection_: LoadKind);
+    }
+
+    pub fn may_start_reporting(&self) {
+        self.reporter.may_start()
+    }
+
+    pub fn may_start_monitoring(&self) {
+        self.monitoring_server.may_start()
+    }
+}
+
+pub struct DiagnosticsResultTracker<'d> {
+    diagnostics_manager: &'d DiagnosticsManager,
+    database_name: Option<&'d str>,
+    action_kind: ActionKind,
+    error_code: Option<String>,
+}
+
+impl<'d> DiagnosticsResultTracker<'d> {
+    pub fn without_database(diagnostics_manager: &'d DiagnosticsManager, action_kind: ActionKind) -> Self {
+        DiagnosticsResultTracker { diagnostics_manager, database_name: None, action_kind, error_code: None }
+    }
+
+    pub fn with_database(
+        diagnostics_manager: &'d DiagnosticsManager,
+        database_name: &'d str,
+        action_kind: ActionKind,
+    ) -> Self {
+        DiagnosticsResultTracker {
+            diagnostics_manager,
+            database_name: Some(database_name),
+            action_kind,
+            error_code: None,
+        }
+    }
+
+    pub fn set_error_code(&mut self, error_code: String) {
+        self.error_code = Some(error_code);
+    }
+}
+
+impl<'d> Drop for DiagnosticsResultTracker<'d> {
+    fn drop(&mut self) {
+        match &self.error_code {
+            Some(error_code) => {
+                self.diagnostics_manager.submit_action_fail(self.database_name, self.action_kind);
+                self.diagnostics_manager.submit_error(self.database_name, error_code.clone());
+            }
+            None => self.diagnostics_manager.submit_action_success(self.database_name, self.action_kind),
+        }
+    }
+}
+
+pub fn run_with_diagnostics<F, T>(
+    diagnostics_manager: &DiagnosticsManager,
+    action_kind: ActionKind,
+    f: F,
+) -> Result<T, Status>
+where
+    F: FnOnce() -> Result<T, Status>,
+{
+    run_and_track(DiagnosticsResultTracker::without_database(diagnostics_manager, action_kind), f)
+}
+
+pub fn run_with_diagnostics_for_database<F, T>(
+    diagnostics_manager: &DiagnosticsManager,
+    database_name: &str,
+    action_kind: ActionKind,
+    f: F,
+) -> Result<T, Status>
+where
+    F: FnOnce() -> Result<T, Status>,
+{
+    run_and_track(DiagnosticsResultTracker::with_database(diagnostics_manager, database_name, action_kind), f)
+}
+
+fn run_and_track<F, T>(mut tracker: DiagnosticsResultTracker<'_>, f: F) -> Result<T, Status>
+where
+    F: FnOnce() -> Result<T, Status>,
+{
+    let result = f();
+    if let Err(ref status) = result {
+        tracker.set_error_code(get_status_error_code(status));
+    }
```

Comment:
I'm not convinced this needs the tracker. `ActionKind` and `Option<Database>` can be arguments to `run_and_track`, and the `drop` glue can be inlined here.

Reply:
Well, you're right, it ended up being quite simple. Cool, I'll remove the tracker.